File: src/worker/webhook.ts
import axios from 'axios';
import crypto from 'crypto';
import { Task } from '../types';
import { config } from '../config';
import { markWebhookDelivered, addTaskLog } from '../db/taskRepository';
import { logger } from '../utils/logger';

export async function deliverWebhook(task: Task): Promise<void> {
  if (!task.webhook_url || task.webhook_delivered) return;

  const payload = {
    task_id: task.id,
    type: task.type,
    status: task.status,
    result: task.result,
    error_message: task.error_message,
    completed_at: task.completed_at,
  };

  const body = JSON.stringify(payload);
  const signature = crypto
    .createHmac('sha256', config.webhook.secret)
    .update(body)
    .digest('hex');

  try {
    await axios.post(task.webhook_url, payload, {
      timeout: config.webhook.timeoutMs,
      headers: {
        'Content-Type': 'application/json',
        'X-JobQueue-Signature': `sha256=${signature}`,
        'X-JobQueue-TaskId': task.id,
        'X-JobQueue-Event': 'task.completed',
      },
    });

    await markWebhookDelivered(task.id);
    await addTaskLog({
      task_id: task.id,
      event: 'webhook_delivered',
      message: `Webhook delivered to ${task.webhook_url}`,
    });
    logger.info('Webhook', `Delivered for task ${task.id} → ${task.webhook_url}`);
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    logger.warn('Webhook', `Failed for task ${task.id}: ${msg}`);
    await addTaskLog({
      task_id: task.id,
      event: 'webhook_failed',
      message: `Webhook delivery failed: ${msg}`,
      metadata: { url: task.webhook_url },
    });
  }
}

File: app/worker/webhook.py
import hmac
import hashlib
import json
import urllib.request
import urllib.error
from app.core.config import get_settings
from app.core.logging import logger

settings = get_settings()


def deliver_webhook(task_id: str, task_type: str, status: str,
                    result: dict | None, error_message: str | None,
                    completed_at: str | None, webhook_url: str) -> bool:
    payload = {
        "task_id": task_id,
        "type": task_type,
        "status": status,
        "result": result,
        "error_message": error_message,
        "completed_at": completed_at,
    }
    body = json.dumps(payload).encode()
    signature = hmac.new(
        settings.WEBHOOK_SECRET.encode(), body, hashlib.sha256
    ).hexdigest()

    req = urllib.request.Request(
        webhook_url,
        data=body,
        headers={
            "Content-Type": "application/json",
            "X-JobQueue-Signature": f"sha256={signature}",
            "X-JobQueue-TaskId": task_id,
            "X-JobQueue-Event": "task.completed",
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=settings.WEBHOOK_TIMEOUT):
            logger.info("webhook_delivered", task_id=task_id, url=webhook_url)
            return True
    except (urllib.error.URLError, Exception) as exc:
        logger.warning("webhook_failed", task_id=task_id, url=webhook_url, error=str(exc))
        return False
